### PR TITLE
Aggregate signatures sovereign outgoing tx data

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Added possibility to aggregate sig and add leader sig on extra messages/data and send them through consensus messages
  
## Proposed changes
- An extra signer is assigned in the **end** sub round 
- Each extra signer is registered in a `holder` of signers for the specific sub round
- Each sub round contains the above `holder` of signers to call when necessary. Basically you can add any number of extra custom signatures.
- For each signer, a shallow clone( **NOT** the real object from core compoments) needs to be provided, since it holds internal data of its own signatures

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

File: testscommon/subRounds/subRoundStartExtraSignersHolderMock.go [hunk 1]
```diff
@@ -16,8 +16,8 @@ func (mock *SubRoundStartExtraSignersHolderMock) Reset(pubKeys []string) error {
 	return nil
 }
 
-// RegisterExtraSingingHandler -
-func (mock *SubRoundStartExtraSignersHolderMock) RegisterExtraSingingHandler(extraSigner consensus.SubRoundStartExtraSignatureHandler) error {
+// RegisterExtraSigningHandler -
+func (mock *SubRoundStartExtraSignersHolderMock) RegisterExtraSigningHandler(extraSigner consensus.SubRoundStartExtraSignatureHandler) error {
```
Comment (raduchis) [inline testscommon/subRounds/subRoundStartExtraSignersHolderMock.go:20]: there still are some `Singing` instead of `Signing` - found 8 occurences

**Review by raduchis — CHANGES_REQUESTED** [review]

File: consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// AggregateSignatures aggregates signatures for outgoing tx data
+func (sr *sovereignSubRoundEndOutGoingTxData) AggregateSignatures(bitmap []byte, epoch uint32) ([]byte, error) {
+	sig, err := sr.signingHandler.AggregateSigs(bitmap, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	err = sr.signingHandler.SetAggregatedSig(sig)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go:52]: Right, refactored

File: consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	err = outGoingMb.SetLeaderSignatureOutGoingOperations(leaderSig)
+	if err != nil {
+		return err
+	}
+
+	return sovHeader.SetOutGoingMiniBlockHeaderHandler(outGoingMb)
+}
+
+// HaveConsensusHeaderWithFullInfo sets aggregated and leader signature in header with provided data from consensus message
+func (sr *sovereignSubRoundEndOutGoingTxData) HaveConsensusHeaderWithFullInfo(header data.HeaderHandler, cnsMsg *consensus.Message) error {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go:110]: Named it to be consistent with the func from which this wall called from.
Renamed to `SetConsensusDataInHeader`

File: consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+// SetAggregatedSignatureInHeader sets aggregated signature for outgoing tx in header
+func (sr *sovereignSubRoundEndOutGoingTxData) SetAggregatedSignatureInHeader(header data.HeaderHandler, aggregatedSig []byte) error {
+	sovHeader, castOk := header.(data.SovereignChainHeaderHandler)
+	if !castOk {
+		return fmt.Errorf("%w in sovereignSubRoundEndOutGoingTxData.SetAggregatedSignatureInHeader", errors.ErrWrongTypeAssertion)
+	}
+
+	outGoingMb := sovHeader.GetOutGoingMiniBlockHeaderHandler()
+	if check.IfNil(outGoingMb) {
+		return nil
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply consensus/spos/bls/sovereignSubRoundEndOutGoingTxData.go:69]: This is not the case here.
If you have no outgoing operations, then there is no outgoing mb header => nothing to set => nothing to sign

File: consensus/spos/bls/blsSubroundsFactory.go [hunk 5]
```diff
@@ -131,8 +130,18 @@ func (fct *factory) GenerateSubrounds() error {
 	if err != nil {
 		return err
 	}
+	err = signRoundExtraSignersHolder.RegisterExtraSigningHandler(signRoundExtraSigner)
+	if err != nil {
+		return err
+	}
 
-	err = signRoundExtraSignersHolder.RegisterExtraSingingHandler(signRoundExtraSigner)
+	err = signRoundExtraSignersHolder.RegisterExtraSigningHandler(signRoundExtraSigner)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply consensus/spos/bls/blsSubroundsFactory.go:138]: Refaactor done here: https://github.com/multiversx/mx-chain-go/pull/5717 to only register extra signers in sovereign main file.

File: consensus/spos/bls/subroundEndRound.go [hunk 6]
```diff
@@ -31,20 +40,25 @@ func NewSubroundEndRound(
...
 	displayStatistics func(),
+	extraSignersHolder SubRoundEndExtraSignersHolder,
 ) (*subroundEndRound, error) {
 	err := checkNewSubroundEndRoundParams(
 		baseSubround,
 	)
 	if err != nil {
 		return nil, err
 	}
+	if check.IfNil(extraSignersHolder) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply consensus/spos/bls/subroundEndRound.go:51]: No, that's for every compoment(regular/sovereign). That's the reason I used a `holder`.
For regular chains, the holder will be empty (no registered extra signer) => nothing extra to be signed.

File: testscommon/subRounds/subRoundStartExtraSignersHolderMock.go (hunk 1 above)
Comment (mariusmihaic) [reply testscommon/subRounds/subRoundStartExtraSignersHolderMock.go:20]: Right, changed all remaining ones

**Review by raduchis — DISMISSED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by raduchis — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**